kategory-test/src/main/kotlin/kategory/laws/OptionalLaws.kt

```diff
@@ -0,0 +1,98 @@
+            Law("Optional Law: consistent getOption with modify identity", { consistentGetOptionModifyId(optional, aGen, EQB) })
+    )
+
+    fun <A, B> getOptionSet(optional: Optional<A, B>, aGen: Gen<A>, EQA: Eq<A>): Unit =
+            forAll(aGen, { a ->
+                EQA.eqv(
+                        optional.getOrModify(a).fold(::identity, { optional.set(it)(a) }),
+                        a
+                )
+            })
+
+    fun <A, B> getGetOption(optional: Optional<A, B>, aGen: Gen<A>, bGen: Gen<B>, EQB: Eq<B>): Unit =
+            forAll(aGen, bGen, { a, b ->
+                optional.getOption(optional.set(b)(a)).exists { EQB.eqv(it, b) }
+            })
+
+    fun <A, B> setIdempotent(optional: Optional<A, B>, aGen: Gen<A>, bGen: Gen<B>, EQA: Eq<A>): Unit =
+            forAll(aGen, bGen, { a, b ->
+                EQA.eqv(
+                        optional.set(b)(optional.set(b)(a)),
+                        optional.set(b)(a)
+                )
+            })
+
+    fun <A, B> modifyIdentity(optional: Optional<A, B>, aGen: Gen<A>, EQA: Eq<A>): Unit =
+            forAll(aGen, { a ->
+                EQA.eqv(
+                        optional.modify(::identity)(a),
+                        a
+                )
+            })
+
+
+    fun <A, B> composeModify(optional: Optional<A, B>, aGen: Gen<A>, funcGen: Gen<(B) -> B>, EQA: Eq<A>): Unit =
+            forAll(aGen, funcGen, funcGen, { a, f, g ->
+                EQA.eqv(
+                        optional.modify(g)(optional.modify(f)(a)),
+                        optional.modify(g compose f)(a)
+                )
+            })
+
+    fun <A, B> consistentSetModify(optional: Optional<A, B>, aGen: Gen<A>, bGen: Gen<B>, EQA: Eq<A>): Unit =
+            forAll(aGen, bGen, { a, b ->
+                EQA.eqv(
+                        optional.set(b)(a),
+                        optional.modify { b }(a)
+                )
+            })
+
+    fun <A, B> consistentModifyModifyId(optional: Optional<A, B>, aGen: Gen<A>, funcGen: Gen<(B) -> B>, EQA: Eq<A>): Unit =
+            forAll(aGen, funcGen, { a, f ->
+                EQA.eqv(
+                        optional.modify(f)(a),
+                        optional.modifyF(Id.applicative(), { Id.pure(f(it)) }, a).value()
+                )
+            })
+
+    inline fun <reified A, B> consistentGetOptionModifyId(optional: Optional<A, B>, aGen: Gen<A>, EQB: Eq<B>): Unit =
+            forAll(aGen, { a ->
+                val getOption = optional.getOption(a)
+
+                val modifyFId = optional.modifyF(Const.applicative(firstOptionMonoid<B>()), { b ->
+                    Const(FirstOption(b.some()))
+                }, a).value().option
+
+                getOption.exists { b ->
+                    modifyFId.exists {
+                        EQB.eqv(b, it)
+                    }
+                }
+            })
+
+    @PublishedApi internal data class FirstOption<A>(val option: Option<A>)
```

Comment:
What does @PublishedApi do? It's the first time I see it 😖

Reply:
Oh is this to allow inlines to use internals? Fuck yeah, I've been looking for something like this forever. We can hide some mechanisms now :D